typed-function.js

```diff
@@ -1259,96 +1259,137 @@
-          for (var signature in fn.signatures) {
-            if (fn.signatures.hasOwnProperty(signature)) {
-              validateUnique(signature, fn.signatures[signature]);
-              signaturesMap[signature] = fn.signatures[signature];
+    /**
+     * Copy all of the signatures from the second argument into the first,
+     * which is modified by side effect, checking for conflicts
+     */
+    function mergeSignatures (dest, source) {
+      for (const key in source) {
+        if (source.hasOwnProperty(key)) {
+          if (key in dest) {
+            if (source[key] !== dest[key]) {
+              const err = new Error('Signature "' + key + '" is defined twice')
+              err.data = { signature: key }
+              throw err
             }
+            // else: both signatures point to the same function, that's fine
           }
-        }
-        else if (typeof fn.signature === 'string') {
-          validateUnique(fn.signature, fn);
-          signaturesMap[fn.signature] = fn;
-        }
-        else {
-          err = new TypeError('Function is no typed-function (index: ' + i + ')');
-          err.data = {index: i};
-          throw err;
+          dest[key] = source[key]
         }
       }
-
-      return signaturesMap;
     }
 
-    typed = createTypedFunction('typed', {
-      'string, Object': createTypedFunction,
-      'Object': function (signaturesMap) {
-        // find existing name
-        var fns = [];
-        for (var signature in signaturesMap) {
-          if (signaturesMap.hasOwnProperty(signature)) {
-            fns.push(signaturesMap[signature]);
+    /**
+     * Originally the main function was a typed function itself, but then
+     * it might not be able to generate error messages if the client
+     * replaced the type system with different names.
+     *
+     * Main entry: typed([name], functions/objects with signatures...)
+     *
+     * Assembles and returns a new typed-function from the given items
+     * that provide signatures and implementations, each of which may be
+     * * a plain object mapping (string) signatures to implementing functions,
+     * * a previously constructed typed function, or
+     * * any other single function with a string-valued property `signature`.
+
+     * The name of the resulting typed-function will be given by the
+     * string-valued name argument if present, or if not, by the name
+     * of any of the arguments that have one, as long as any that do are
+     * consistent with each other. If no name is specified, the name will be
+     * an empty string.
+     *
+     * @param {string} name [optional]
+     * @param {(function|object)[]} signature providers
+     * @returns {typed-function}
+     */
+    typed = function(maybeName) {
+      const named = typeof maybeName === 'string'
+      let name = named ? maybeName : ''
+      let start = named ? 1 : 0
+      let allSignatures = {}
+      for (let i = start; i < arguments.length; ++i) {
```

Comment:
This for-loop is quite large and has quite some nesting. At first sight it looks complex. Open question: are there ways to make this function smaller? Some ideas:

- extract checks into util functions like `isPlainObject`, `isTypedFunction`, `isFunctionWithSignature` and use them here.
- flatten the cases 1, 2, 3, and only have one error handling case at the end (instead of the current two)

Reply:
Yeah, I thought you might be concerned about the number of lines of code in this function. I will try to streamline/factor it, thanks for the suggestions about possible tactics for it.

Reply:
OK, how's the version I just committed? Once some version of #135 is merged, I can change case 2 to use isTypedFunction; I didn't want to make this PR and that one overlap in a way that might make merging/rebasing tricky.

Reply:
Nice, thanks! I like the result, it is easier to follow now 👍